File: server/Schema/SchemaItem.cs
//////////////////////////////////////////////////////////////////////////
// 
// 文件：Schema/SchemaItem.cs
// 作者：Xoen Xie
// 时间：2023/6/19
// 描述：物品配置
// 说明：
//
//////////////////////////////////////////////////////////////////////////
#pragma warning disable CS8600, CS8602, CS8603, CS8618, CS8601
using XSF;

namespace XsfScp
{
    public class SchemaItem : ISchema
    {
        private Dictionary<uint, ScpItem> m_Datas;

        public string GetSchemaName(string name)
        {
            return name;
        }

        public void OnSchemaLoad(ISchemaReader reader)
        {
            m_Datas = new Dictionary<uint, ScpItem>();
            CSVReader csv = reader as CSVReader;

            for (int i = 0; i < csv.mRowCount; ++i)
            {
                ScpItem scp = new ScpItem();

    //_CSV_LIST_BEGIN_
				scp.uId = (csv.GetData((int)CSVDataType.Uint, i, (int)CSVIndex.ScpItem_id) as CSVData_Uint).uValue;	// 编号
				scp.iType = (csv.GetData((int)CSVDataType.Int, i, (int)CSVIndex.ScpItem_type) as CSVData_Int).iValue;	// 类型
				scp.fParam1 = (csv.GetData((int)CSVDataType.Float, i, (int)CSVIndex.ScpItem_param1) as CSVData_Float).fValue;	// 参数1
				scp.ulParam2 = (csv.GetData((int)CSVDataType.Ulong, i, (int)CSVIndex.ScpItem_param2) as CSVData_Ulong).ulValue;	// 参数2
				scp.arParam3 = (csv.GetData((int)CSVDataType.Array, i, (int)CSVIndex.ScpItem_param3) as CSVData_Array).arValue;	// 参数3
				scp.icParam4 = (csv.GetData((int)CSVDataType.IDAndCount, i, (int)CSVIndex.ScpItem_param4) as CSVData_IDAndCount).icValue;	// 参数4
				scp.bParam5 = (csv.GetData((int)CSVDataType.Bool, i, (int)CSVIndex.ScpItem_param5) as CSVData_Bool).bValue;	// 参数5
    //_CSV_LIST_END_

                //* 使用字典
                if (m_Datas.ContainsKey(scp.uId))
                {
                    throw new XSFSchemaLoadException("SchemaItem.OnSchemaLoad key exist, id=" + scp.uId);
                }

                m_Datas.Add(scp.uId, scp);
                //*/
            }
        }

        //* 使用字典 根据ID获取
        public ScpItem Get(uint nID)
        {
            ScpItem scp = null;
            m_Datas.TryGetValue(nID, out scp);
            return scp;
        }
        //*/
    }
}

File: server-cpp/source/schema/src/schemas/SchemaItem.cpp
//////////////////////////////////////////////////////////////////////////
//
// 文件：schema\src\schemas\SchemaItem.cpp
// 作者：Xoen Xie
// 时间：11/10/2023
// 描述：物品配置
// 说明：
//
//////////////////////////////////////////////////////////////////////////
#include "SchemaItem.h"
#include "SchemaIndex.h"

#include "XSF.h"
#include "CSVDataInc.h"

namespace xsf_scp
{
    bool SchemaItem::OnSchemaLoad(int32 nSchemaID, ISchemaReader *pReader)
    {
        for(uint32 i = 0; i < pReader->RowCount(); i ++)
        {
            ScpItem * scp = new ScpItem();
            //_CSV_LIST_BEGIN_
			scp->uId = GET_DATA(CSVData_Uint, CSVDataType_Uint, i, CSVIndex_ScpItem_id)->uValue;;	// 编号
			scp->iType = GET_DATA(CSVData_Int, CSVDataType_Int, i, CSVIndex_ScpItem_type)->iValue;;	// 类型
			scp->fParam1 = GET_DATA(CSVData_Float, CSVDataType_Float, i, CSVIndex_ScpItem_param1)->fValue;;	// 参数1
			scp->ulParam2 = GET_DATA(CSVData_Ulong, CSVDataType_Ulong, i, CSVIndex_ScpItem_param2)->ulValue;;	// 参数2
			scp->arParam3 = GET_DATA(CSVData_Array, CSVDataType_Array, i, CSVIndex_ScpItem_param3)->arValue;;	// 参数3
			scp->icParam4 = GET_DATA(CSVData_IDAndCount, CSVDataType_IDAndCount, i, CSVIndex_ScpItem_param4)->icValue;;	// 参数4
			scp->bParam5 = GET_DATA(CSVData_Bool, CSVDataType_Bool, i, CSVIndex_ScpItem_param5)->bValue;;	// 参数5
            //_CSV_LIST_END_

            if(!(m_Datas.insert(DataMap::value_type(scp->uId, scp)).second))
            {
                XSF_ERROR("SchemaItem::OnSchemaLoad id exist, id=%u", scp->uId);
                return false;
            }
        }

        return true;
    }

    void *SchemaItem::Get(uint32 nID1, uint32 nID2)
    {
        auto it = m_Datas.find(nID1);
        if(it != m_Datas.end())
            return it->second;
    
        return nullptr;
    }
}
